pkg/deploy/controller/deployment/controller.go

```diff
@@ -141,14 +136,34 @@ func (c *DeploymentController) Handle(deployment *kapi.ReplicationController) er
 			glog.V(4).Infof("Cancelling %d deployer pods for deployment %s", len(deployerPods), deployutil.LabelForDeployment(deployment))
 			zeroDelay := int64(1)
 			for _, deployerPod := range deployerPods {
-				// Set the ActiveDeadlineSeconds on the pod so it's terminated very soon.
-				if deployerPod.Spec.ActiveDeadlineSeconds == nil || *deployerPod.Spec.ActiveDeadlineSeconds != zeroDelay {
-					deployerPod.Spec.ActiveDeadlineSeconds = &zeroDelay
-					if _, err := c.podClient.updatePod(deployerPod.Namespace, &deployerPod); err != nil {
-						c.recorder.Eventf(deployment, "failedCancellation", "Error cancelling deployer pod %s for deployment %s: %v", deployerPod.Name, deployutil.LabelForDeployment(deployment), err)
-						return fmt.Errorf("couldn't cancel deployer pod %s for deployment %s: %v", deployerPod.Name, deployutil.LabelForDeployment(deployment), err)
+				if deployerPod.Status.Phase == kapi.PodPending {
+					// If the pod hasn't yet been scheduled, just delete it.
+					err := c.podClient.deletePod(deployerPod.Namespace, deployerPod.Name)
+					if err != nil {
+						c.recorder.Eventf(deployment, "failedCancellation", "Error deleting deployer pod %s while cancelling deployment %s: %v", deployerPod.Name, deployutil.LabelForDeployment(deployment), err)
+						return fmt.Errorf("couldn't delete deployer pod %s while cancelling deployment %s: %v", deployerPod.Name, deployutil.LabelForDeployment(deployment), err)
+					}
+					// If we deleted the unscheduled primary deployer pod, just
+					// transition to failed immediately, since the DeployerPodController
+					// has no cleanup to do as the deployment never executed.
+					if deployerPod.Name == deployutil.DeployerPodNameForDeployment(deployment.Name) {
```

Comment:
On Aug 27, 2015, at 1:45 PM, Dan Mace <notifications@github.com> wrote:

In pkg/deploy/controller/deployment/controller.go
<https://github.com/openshift/origin/pull/4364#discussion_r38117582>:

> -				if deployerPod.Spec.ActiveDeadlineSeconds == nil || *deployerPod.Spec.ActiveDeadlineSeconds != zeroDelay {
> -					deployerPod.Spec.ActiveDeadlineSeconds = &zeroDelay
> -					if _, err := c.podClient.updatePod(deployerPod.Namespace, &deployerPod); err != nil {
> -						c.recorder.Eventf(deployment, "failedCancellation", "Error cancelling deployer pod %s for deployment %s: %v", deployerPod.Name, deployutil.LabelForDeployment(deployment), err)
> -						return fmt.Errorf("couldn't cancel deployer pod %s for deployment %s: %v", deployerPod.Name, deployutil.LabelForDeployment(deployment), err)
> +				if deployerPod.Status.Phase == kapi.PodPending {
> +					// If the pod hasn't yet been scheduled, just delete it.
> +					err := c.podClient.deletePod(deployerPod.Namespace, deployerPod.Name)
> +					if err != nil {
> +						c.recorder.Eventf(deployment, "failedCancellation", "Error deleting deployer pod %s while cancelling deployment %s: %v", deployerPod.Name, deployutil.LabelForDeployment(deployment), err)
> +						return fmt.Errorf("couldn't delete deployer pod %s while cancelling deployment %s: %v", deployerPod.Name, deployutil.LabelForDeployment(deployment), err)
> +					}
> +					// If we deleted the unscheduled primary deployer pod, just
> +					// transition to failed immediately, since the DeployerPodController
> +					// has no cleanup to do as the deployment never executed.
> +					if deployerPod.Name == deployutil.DeployerPodNameForDeployment(deployment.Name) {

I'm not sure we can avoid races unless we can delete w/ a versioned
resource and handle conflicts on delete. Then we could do:

   1. Set deadline on deployer pod
   2. If deployer pod has empty node IP, try to delete
      1. If delete fails with conflict, pod might now be scheduled, ignore-
      deadline/deployerpodcontroller takes over
      2. If delete succeeds, you know it was deleted w/ an empty node IP
      and we can fail the deployment

Without conflicts returned for deletes of a stale resource, we'll always be
racing the scheduler no matter what we check before calling delete, right?



Why do we even use active deadline seconds again?  Why aren't we just
deleting the pod?  Is it just to get logs?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4364/files#r38117582>.